services/supervisor/src/lib.rs

```diff
@@ -97,23 +109,64 @@ fn update_configs(context: &mut CallContext<'_>, changes: Vec<ConfigChange>) {
                 );
             }
         }
+
+        ConfigChange::StartService(start_service) => {
+            log::trace!(
+                "Request add service with name {:?} from artifact {:?}",
+                start_service.name,
+                start_service.artifact
+            );
+            if let Err(e) = context.start_adding_service(
+                start_service.artifact,
+                start_service.name,
+                start_service.config,
+            ) {
+                // Panic will cause changes to be rolled back.
+                panic!(
+                    "An error occurred while starting the service instance: {:?}",
+                    e
+                );
+            }
+        }
     })
 }
 
-#[derive(Debug, ServiceFactory)]
+#[derive(Debug, Default, Clone, ServiceFactory)]
 #[exonum(
     proto_sources = "proto",
     artifact_name = "exonum-supervisor",
+    service_constructor = "Self::construct",
     implements("transactions::SupervisorInterface")
 )]
-pub struct Supervisor;
+pub struct Supervisor<Mode>
```

Comment:
Not sure about this. With current approach, `SimpleSupervisor` and `DecentralizedSupervisor` are different types which are harder to confuse than two `Supervisor`s with different contents.
Once `const` generics are stabilized, we'll be able to use enum variants here as well instead of different structs.

Reply:
> which are harder to confuse than two Supervisors with different contents.

This assumes that **(a)** validator constraints cannot change during blockchain operation; **(b)** it's easier to look at the source code of the node than at its genesis block or HTTP API (provided supervisor config is returned from there). I'm having serious doubts about either of these assumptions.

Reply:
Regarding **a**: hm, do you mean that supervisor mode should be changeable part of configuration?
Regarding **b**: it worth nothing to implement API endpoint showing the kind of supervisor, or even have different instance names.

To be clear:
My motivation was based on the assumtion that `SimpleSupervisor` and `DecentralizedSupervisor` are not interchangeable, decentralized supervisor cannot suddenly become a simple one and vice versa. If they will be considered a volatile part of config, then I agree with you, a provided value will be much better.